overlord/ifacestate/handlers.go

```diff
@@ -485,6 +487,143 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
+		// snap and the new core snap providing the same interface. In that
+		// situation we want to ignore any candidates in ubuntu-core and simply
+		// go with those from the new core snap.
+		if len(candidates) == 2 {
+			switch {
+			case candidates[0].Snap.Name() == "ubuntu-core" && candidates[1].Snap.Name() == "core":
+				candidates = candidates[1:2]
+			case candidates[1].Snap.Name() == "ubuntu-core" && candidates[0].Snap.Name() == "core":
+				candidates = candidates[0:1]
+			}
+		}
+		if len(candidates) != 1 {
+			crefs := make([]string, 0, len(candidates))
+			for _, candidate := range candidates {
+				crefs = append(crefs, candidate.String())
+			}
+			task.Logf("cannot auto connect %s (plug auto-connection), candidates found: %q", plug, strings.Join(crefs, ", "))
+			continue
+		}
+		slot := candidates[0]
+		connRef := interfaces.NewConnRef(plug, slot)
+		key := connRef.ID()
+		if _, ok := conns[key]; ok {
+			// Suggested connection already exist so don't clobber it.
+			// NOTE: we don't log anything here as this is a normal and common condition.
+			continue
+		}
+
+		ts, err := AutoConnect(st, plug.Snap.Name(), plug.Name, slot.Snap.Name(), slot.Name)
+		if err != nil {
+			task.Logf("cannot auto connect %s to %s: %s (plug auto-connection)", connRef.PlugRef, connRef.SlotRef, err)
+			continue
+		}
+		autoconnectTs = append(autoconnectTs, ts)
+	}
+	// Auto-connect all the slots
+	for _, slot := range m.repo.Slots(snapName) {
+		candidates := m.repo.AutoConnectCandidatePlugs(snapName, slot.Name, autochecker.check)
+		if len(candidates) == 0 {
+			continue
+		}
+
+		for _, plug := range candidates {
+			// make sure slot is the only viable
+			// connection for plug, same check as if we were
+			// considering auto-connections from plug
+			candSlots := m.repo.AutoConnectCandidateSlots(plug.Snap.Name(), plug.Name, autochecker.check)
+
+			if len(candSlots) != 1 || candSlots[0].String() != slot.String() {
+				crefs := make([]string, 0, len(candSlots))
+				for _, candidate := range candSlots {
+					crefs = append(crefs, candidate.String())
+				}
+				task.Logf("cannot auto connect %s to %s (slot auto-connection), alternatives found: %q", slot, plug, strings.Join(crefs, ", "))
+				continue
+			}
+
+			connRef := interfaces.NewConnRef(plug, slot)
+			key := connRef.ID()
+			if _, ok := conns[key]; ok {
+				// Suggested connection already exist so don't clobber it.
+				// NOTE: we don't log anything here as this is a normal and common condition.
+				continue
+			}
+			ts, err := AutoConnect(st, plug.Snap.Name(), plug.Name, slot.Snap.Name(), slot.Name)
+			if err != nil {
+				task.Logf("cannot auto connect %s to %s: %s (slot auto-connection)", connRef.PlugRef, connRef.SlotRef, err)
+				continue
+			}
+			autoconnectTs = append(autoconnectTs, ts)
+		}
+	}
+
+	chg := task.Change()
+	lanes := task.Lanes()
+	if len(lanes) == 1 && lanes[0] == 0 {
+		lanes = nil
+	}
+	ht := task.HaltTasks()
+	for _, ts := range autoconnectTs {
```

Comment:
The logic below looks nice and clear, but note that we're doing the exact same logic to all of the individual task sets, so in effect we're handling it as a single task set. So it would probably be more clear to do exactly that: use a single task set, and then run these operations just once, on the merged task set itself.

In practice, declare the task set before the earlier loop, and instead of _autoconnectTs_, declare something like:
```
var autots state.TaskSet
```
then, instead of _append_, do:
```
autots.AddAll(ts)
``` 
Now we have a single task set, and can just do the logic below once:
```
...etc...
chg.AddAll(autots)
...etc...
```

Reply:
Thanks for the suggestion, indeed it's much simpler, I missed that detail of the state API.